File: OneSignalSDK/onesignal/src/main/java/com/onesignal/BadgeCountUpdater.java
/**
 * Modified MIT License
 *
 * Copyright 2017 OneSignal
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * 2. All copies of substantial portions of the Software may only be used in connection
 * with services provided by OneSignal.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.onesignal;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.os.Build;
import android.os.Bundle;
import android.service.notification.StatusBarNotification;
import android.support.annotation.RequiresApi;

import com.onesignal.OneSignalDbContract.NotificationTable;
import com.onesignal.shortcutbadger.ShortcutBadgeException;
import com.onesignal.shortcutbadger.ShortcutBadger;

import static com.onesignal.NotificationLimitManager.MAX_NUMBER_OF_NOTIFICATIONS_STR;

class BadgeCountUpdater {

   // Cache for manifest setting.
   private static int badgesEnabled = -1;

   private static boolean areBadgeSettingsEnabled(Context context) {
      if (badgesEnabled != -1)
         return (badgesEnabled == 1);

      try {
         ApplicationInfo ai = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
         Bundle bundle = ai.metaData;
         if (bundle != null) {
            String defaultStr = bundle.getString("com.onesignal.BadgeCount");
            badgesEnabled = "DISABLE".equals(defaultStr) ? 0 : 1;
         }
         else
            badgesEnabled = 1;
      } catch (PackageManager.NameNotFoundException e) {
         badgesEnabled = 0;
         OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Error reading meta-data tag 'com.onesignal.BadgeCount'. Disabling badge setting.", e);
      }

      return (badgesEnabled == 1);
   }
   
   private static boolean areBadgesEnabled(Context context) {
      return areBadgeSettingsEnabled(context) && OSUtils.areNotificationsEnabled(context);
   }

   static void update(OneSignalDb db, Context context) {
      if (!areBadgesEnabled(context))
         return;

      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
         updateStandard(context);
      else
         updateFallback(db, context);
   }

   @RequiresApi(api = Build.VERSION_CODES.M)
   private static void updateStandard(Context context) {
      StatusBarNotification[] activeNotifs = OneSignalNotificationManager.getActiveNotifications(context);

      int runningCount = 0;
      for (StatusBarNotification activeNotif : activeNotifs) {
         if (NotificationLimitManager.isGroupSummary(activeNotif))
            continue;
         runningCount++;
      }

      updateCount(runningCount, context);
   }

   private static void updateFallback(OneSignalDb db, Context context) {
      Cursor cursor = db.query(
         NotificationTable.TABLE_NAME,
         null,
         OneSignalDbHelper.recentUninteractedWithNotificationsWhere().toString(),
         null,                                                    // Where args
         null,                                                    // group by
         null,                                                    // filter by row groups
         null,                                                     // sort order, new to old
         MAX_NUMBER_OF_NOTIFICATIONS_STR
      );

      int notificationCount = cursor.getCount();
      cursor.close();

      updateCount(notificationCount, context);
   }

   static void updateCount(int count, Context context) {
      if (!areBadgeSettingsEnabled(context))
         return;

      try {
         ShortcutBadger.applyCountOrThrow(context, count);
      } catch (ShortcutBadgeException e) {
         // Suppress error as there are normal cases where this will throw
         // Can throw if:
         //    - Badges are not support on the device.
         //    - App does not have a default launch Activity.
      }
   }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/notification/internal/badges/BadgeCountUpdater.kt
package com.onesignal.onesignal.notification.internal.badges

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.annotation.RequiresApi
import com.onesignal.onesignal.core.internal.database.IDatabaseProvider
import com.onesignal.onesignal.core.internal.database.impl.OneSignalDbContract
import com.onesignal.onesignal.notification.internal.NotificationHelper
import com.onesignal.onesignal.notification.internal.common.INotificationQueryHelper
import com.onesignal.onesignal.notification.internal.generation.NotificationLimitManager
import com.onesignal.onesignal.core.internal.logging.Logging
import com.onesignal.shortcutbadger.ShortcutBadger
import com.onesignal.shortcutbadger.ShortcutBadgeException

internal class BadgeCountUpdater(
    private val _queryHelper: INotificationQueryHelper,
    private val _databaseProvider: IDatabaseProvider
) {
    // Cache for manifest setting.
    private var badgesEnabled = -1
    private fun areBadgeSettingsEnabled(context: Context): Boolean {
        if (badgesEnabled != -1) return badgesEnabled == 1
        try {
            val ai = context.packageManager.getApplicationInfo(
                context.packageName,
                PackageManager.GET_META_DATA
            )
            val bundle = ai.metaData
            if (bundle != null) {
                val defaultStr = bundle.getString("com.onesignal.BadgeCount")
                badgesEnabled = if ("DISABLE" == defaultStr) 0 else 1
            } else badgesEnabled = 1
        } catch (e: PackageManager.NameNotFoundException) {
            badgesEnabled = 0
            Logging.error("Error reading meta-data tag 'com.onesignal.BadgeCount'. Disabling badge setting.", e)
        }
        return badgesEnabled == 1
    }

    private fun areBadgesEnabled(context: Context): Boolean {
        return areBadgeSettingsEnabled(context) && NotificationHelper.areNotificationsEnabled(context)
    }

    fun update(context: Context) {
        if (!areBadgesEnabled(context)) return
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
            updateStandard(context)
        else
            updateFallback(context)
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    private fun updateStandard(context: Context) {
        val activeNotifs = NotificationHelper.getActiveNotifications(context)
        var runningCount = 0
        for (activeNotif in activeNotifs) {
            if (NotificationHelper.isGroupSummary(activeNotif)) continue
            runningCount++
        }
        updateCount(runningCount, context)
    }

    private fun updateFallback(context: Context) {
        var notificationCount: Int = 0

        _databaseProvider.get().query(
            OneSignalDbContract.NotificationTable.TABLE_NAME,
            null,
            _queryHelper.recentUninteractedWithNotificationsWhere().toString(),
            null,  // Where args
            null,  // group by
            null,  // filter by row groups
            null,  // sort order, new to old
            NotificationLimitManager.maxNumberOfNotificationsInt.toString()
        ).use {
            notificationCount = it.count
        }
        updateCount(notificationCount, context)
    }

    fun updateCount(count: Int, context: Context) {
        if (!areBadgeSettingsEnabled(context)) return
        try {
            ShortcutBadger.applyCountOrThrow(context, count)
        } catch (e: ShortcutBadgeException) {
            // Suppress error as there are normal cases where this will throw
            // Can throw if:
            //    - Badges are not support on the device.
            //    - App does not have a default launch Activity.
        }
    }
}